### PR TITLE
Added error sending function and a strategy for adding registering error

### PR DESCRIPTION
**Pull request by thongsav-usgs**
handling consumers per request/serviceRequest.

### CODE REVIEW

File: src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java [hunk 1]
```diff
@@ -6,10 +6,16 @@
```
Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java]: was this class being used before Msgservice started implementing it?

File: src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java [hunk 2]
```diff
@@ -6,10 +6,16 @@
...
+	/**
 	 * 
-	 * @param topic
 	 * @param requestId
-	 * @param message
+	 * @param serviceRequestId
+	 * @param e
 	 */
-	public void sendMessage(String topic, String requestId, Map<String, String> message);
+	public void postError(String requestId, String serviceRequestId, Exception e);
```
Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java]: If we saw the need to shift sendMessage from explicitly using requestId, why are we putting it into postError?

File: src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java (hunk 1 above)
Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java]: Not that I saw, I assumed it was forgotten and I readded it. Nothing in AQCU was using it either.

File: src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java (hunk 2 above)
Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/api/messaging/MessagingClient.java]: I'm not sure that that was an explicit decision, I'll play with a refactor to bring those ids back into the method signature to see if it makes sense.

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java [hunk 3]
```diff
@@ -1,29 +1,35 @@
...
  */
 public class MicroserviceConsumer extends DefaultConsumer  {
 	private static final Logger log = LoggerFactory.getLogger(MicroserviceConsumer.class);
 
 	private final Channel channel;
 	private final MicroserviceHandler handler;
+	private final MicroserviceMsgservice msgService;
 	
-	public MicroserviceConsumer(Channel channel, MicroserviceHandler handler) {
+	public MicroserviceConsumer(Channel channel, MicroserviceHandler handler, MicroserviceMsgservice msgService) {
```
Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java]: unless an injection strategy requires it, what's the point of having msgService in the constructor?  Will there be multiple implementations to choose from?  Rather, we may instead want to replace this with something like an ErrorHandler, for which Msgservice can provide a default.

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java [hunk 4]
```diff
@@ -40,9 +46,18 @@ public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProp
...
-			log.error("handleDelivery", e);
+			log.error("Error while handling message", e);
+			msgService.postError(
+					getStringFromHeaders(properties.getHeaders(), "requestId"),
+					getStringFromHeaders(properties.getHeaders(), "serviceRequestId"), 
+					e);
 		}
 	}
+
+	private String getStringFromHeaders(Map<String, Object> params, String key) {
```
Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java]: forgot to pull this out when you added MessageUtils.getStringFromHeaders

Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java]: The consumers actually catch exceptions and is the only thing that calls postError. It needs a service instance to be able to call postError.

All application code which runs from MicroserviceConsumers only have to throw exceptions, they don't know anything about error messaging.

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java (hunk 3 above)
Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceConsumer.java]: The consumers actually catch exceptions and is the only thing that calls postError. It needs a service instance to be able to call postError.

All application code which runs from MicroserviceConsumers only have to throw exceptions, they don't know anything about error messaging.

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java [hunk 5]
```diff
@@ -189,6 +196,89 @@ public MicroserviceMsgservice(String host, String exchange, String username, Str
...
 	}
+	
+	/**
+	 * TODO
+	 * @param requestId
+	 * @param serviceRequestId
+	 * @param exceptionClass
+	 * @param errorHandlerClass
+	 */
+	public void registerErrorHandler(String requestId, String serviceRequestId, Class<? extends Exception> exceptionClass, Class<? extends MicroserviceErrorHandler> errorHandlerClass) {
```
Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: I'm confused at how you'd use this.  would every microservice war instance be setting up bindings to listen to every serviceRequestId it ever operates on?

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java [hunk 6]
```diff
@@ -59,6 +64,8 @@
 	public static MicroserviceMsgservice INSTANCE = null;
 	public static String SERVICE_NAME = null;
 	public static Set<Class<? extends MicroserviceHandler>> HANDLERS = Collections.EMPTY_SET;
+	
+	public static final String ERROR_QUEUE_NAME = "aqcu-errors";
```
Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: anything aqcu namespaced probably shouldn't be in the microservices library

Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: Doh, good catch.

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java (hunk 5 above)
Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: I was thinking that you call register in the same way you'd wrap something in a try-catch. The current pattern ends up registering a lot of handlers at runtime (note the TODO further down).

I'm pretty sure this is effectively a memory leak.

Comment (davidmsibley) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: yeah, so the workflow confuses me then.  Bindings are not the types of messages this consumer gets, but the types of messages that are routed to the queue.  So it looks like we're mapping a bunch of very specific bindings, but also creating a bunch of consumers on that one queue that seemingly could be operating on errors outside of its domain.

**Comment by davidmsibley** [comment]
So my overhead view of this is adding a message publish of the exception inside that catch statement in the consumer makes sense, but I don't understand adding the rest of this stuff down this low. The postError method is just a sendMessage wrapper, and the MicroserviceErrorHandler lifecycle is confusing.  I think that stuff would make more sense in it's own microservice.

**Comment by thongsav-usgs** [comment]
I intended MicroserviceErrorHandler to be kind of a callback. It's how end users of the MsgServices provide handlers per request/service request.

**Comment by thongsav-usgs** [comment]
See LookupOrchestrator.java in https://github.com/USGS-CIDA/AQCU/compare/master...thongsav-usgs:error-handling?expand=1 to see how I was planning on using this.

File: src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java (hunk 5 above)
Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: Right, it's specific bindings to allow a runtime process (which generates request IDs or SRIs) to subscribe to error messages only tied to the IDs/requests it generated. The idea was to handle errors tied to request scope, not domain. I think we could do somain based errors too, but then we'd have to start tracking/creating an error domain.

Comment (thongsav-usgs) [inline src/main/java/gov/usgs/cida/microservices/messaging/MicroserviceMsgservice.java]: Actually, you're right... the dynamic subscription model could be handled using the operator + the normal Handlers/Consumers.

**Comment by thongsav-usgs** [comment]
Pulled out the ErrorHandling stuff (error subscription). That will be up to microservices to handle.

**Comment by davidmsibley** [comment]
Yeah this looks much cleaner

**Pull request merged**